docs/specs/captcha.md

```diff
@@ -1,41 +1,354 @@
+- `captcha.providers.type`: Required. The type of the Captcha provider. Valid values are `cloudflare` and `recaptchav2`.
+- `captcha.providers.alias`: Required. The unique identifier to the Captcha provider. It is used in other parts of the configuration to refer to this particular Captcha provider. For example, the project can configured a number of Captcha providers, but only uses one of them in a particular Authentication Flow.
 
-### authgear.secrets.yaml
+Other fields are specific to `type`.
+
+#### `type: cloudflare`
 
-When the captcha provider is configured, the corresponding secret must exist in `authgear.secrets.yaml`.
+- `site_key`: The site key of Cloudflare Turnstile.
+
+#### `type: recaptchav2`
+
+- `site_key`: The site key of reCAPTCHA v2.
+
+### authgear.secrets.yaml
 
 ```yaml
 - data:
-    secret: YOUR_TURNSTILE_SECRET_KEY
-  key: captcha.cloudflare
+    items:
+    - type: cloudflare
+      alias: cloudflare
+      secret: TURNSTILE_SECRET_KEY
+  key: captcha.providers
+```
+
+- `items.type`: Required. It is the same as `captcha.providers.type`.
+- `items.alias`: Required. It is the same as `captcha.providers.alias`.
+
+Other fields are specific to `type`.
+
+#### `type: cloudflare`
+
+- `secret_key`: Required. The secret key of Cloudflare Turnstile.
+
+#### `type: recaptchav2`
+
+- `secret_key`: Required. The secret key of reCAPTCHA v2.
+
+## Authentication Flow
+
+This section specifies how Captcha works in a Authentication Flow.
+
+### Captcha in Authentication Flow configuration
+
+Captcha is supported in the following flow types:
+
+- `signup`
+- `promote`
+- `login`
+- `signup_login`
+- `reauth`
+- `account_recovery`
+
+Captcha is supported only in the following step types:
+
+- `identify` in `signup`, `promote`, `login`, `signup_login`, and `account_recovery`.
+- `create_authenticator` in `signup` and `promote`.
+- `authenticate` in `login` and `reauth`.
+
+We can see that all supported step types have branches.
+
+To specify Captcha is required,
+
+1. Add `captcha.enabled=true` to the root of the flow configuration.
+2. Add `captcha.required=true` to the branches you want.
+
+For example,
+
+```
+authentication_flow:
+  signup_flows:
+  - name: default
+    captcha:
+      enabled: true
+    steps:
+    - type: identify
+      one_of:
+      - identification: email
+        # Identify with email requires captcha.
+        captcha:
+          required: true
+    - type: authenticate
+      one_of:
+      - authentication: primary_password
+      - authentication: primary_oob_otp_email
 ```
 
-- Provider: `cloudflare`
-  - `key`: `captcha.cloudflare`
-  - `data.secret`: The turnstile secret key
+### Behavior of generated flows
+
+Given `captcha.enabled=true` and `captcha.providers` is non-empty,
+
+1. The generated flow has `captcha.enabled=true` in its root configuration.
+2. All the branches of the first step (that is, the `identify` step) has `captcha.required=true`.
```

Comment:
> All the branches of the first step (that is, the `identify` step) has `captcha.required=true`.

> The identify step in reauth is done by an ID token, which involves no user interaction. So the captcha is supported in authenticate instead.

@louischan-oursky just to confirm, for `reauth` flow, `first step` here would refer to `authenticate`?

Context: adding captcha to `pkg/lib/authenticationflow/declarative/generate_config_reauth_flow.go`, wonder if I should add to `Identify` step or `Authenticate` step 🙏